Jay/scale changes

mritunjaykumar opened the pull request:
Code changes related to load balancing feature.
Earlier, ingestion data was partitioned on tenantId; now, it's partitioned on tenantId+measurement.
This code change doesn't include "rolled-up data ingestion".

Commit:
influxdb scale related changes

Commit:
changes related to schema change

Commit:
changed rest client to InfluxDB java client and related changes

Commit:
adding gitignore file

Commit:
adding metrics to measure influxDB write latency, getting routing info latency and Kafka related metrics.

Commit:
Fixed all raw-ingestion related unit test cases. Rollup-ingestion test cases are still broken. I will fix them, when I am working on rollup.

Commit:
Removed commented code blocks

itzg commented:
@mritunjaykumar , sorry for the delay reviewing this. I am now starting on the review.

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/utils/InfluxDBUtils.java:
```diff
@@ -1,5 +1,8 @@
+import org.influxdb.InfluxDB;
+import org.influxdb.InfluxDBFactory;
+
 public class InfluxDBUtils {
```
Can this class be combined with `InfluxDBHelper` since the suffixes "helper" and "utils" don't make it apparent the difference between these two?

mritunjaykumar replied:
Done

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/InfluxDBHelper.java:
```diff
@@ -115,190 +193,142 @@ private TenantRoutes getTenantRoutes(String tenantId) throws Exception {
+     * @return
+     */
+    private InfluxDB getInfluxDBClient(String instanceUrl) {
+        InfluxDB influxDB = this.urlInfluxDBInstanceMap.get(instanceUrl);
```
I'm thinking you should wrap this in a `synchronized(urlInfluxDBInstanceMap)` block since I'm assuming calls to this originate from within a kafka consumer thread.

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/listeners/processors/TenantIdAndMeasurement.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.RequiredArgsConstructor;
+
+@Data
+@RequiredArgsConstructor
```
Doesn't seem to be needed since you declared a constructor.

mritunjaykumar replied:
Submitted fixes

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/listeners/rawlisteners/RawListener.java:
```diff
@@ -44,50 +56,48 @@ public RawListener(InfluxDBHelper influxDBHelper, MeterRegistry registry) {
             @Header(KafkaHeaders.OFFSET) final long offset,
             final Acknowledgment ack) throws Exception {
 
+        registry.gauge("ingestion.records.count", Arrays.asList(rawListenerTag), gaugeRecordsCount);
```
Could this be registered once in the constructor rather than doing a no-op register here?

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/listeners/rawlisteners/RawListener.java:
```diff
@@ -44,50 +56,48 @@ public RawListener(InfluxDBHelper influxDBHelper, MeterRegistry registry) {
 
-        return processPostInfluxDbIngestion(records.toString(),
-                partitionId, offset, ack, isInfluxDbIngestionSuccessful);
+        processPostInfluxDbIngestion(records.toString(), partitionId, offset, ack);
```
Computing the `toString` of the records collection seems excessive when debug logs are not enabled. You could pass `records` directly and then let slf4j make the optimization here

https://github.com/racker/ceres-ingestion-service/blob/ccc7604d4a83f6829d3545c7586f766dd89e8b93/src/main/java/com/rackspacecloud/metrics/ingestionservice/listeners/UnifiedMetricsListener.java#L37

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/listeners/rawlisteners/deserializer/AvroDeserializer.java:
```diff
@@ -31,7 +31,20 @@ public void configure(Map<String, ?> map, boolean b) { }
+
         String stringifiedData = new String(data);
+
+        if(stringifiedData.contains("\"collectionTarget\":null")) {
```
You can probably remove these content fixes since the maas-metrics-enrich is now normalizing those fields

https://github.com/Rackspace-Segment-Support/maas-metrics-enrich/blob/master/src/main/java/com/rackspace/umb/maasmetricsenrich/services/EnrichmentService.java#L63-L64

mritunjaykumar replied:
Removed it. I also removed `String stringifiedData = new String(data)` and pushed stringification into `debug` and `error` calls. So that stringification only happens in those cases rather than every time which is costly affair.

itzg commented on src/main/resources/avro/ExternalMetric.avsc:
```diff
@@ -1,7 +1,7 @@
 {
```
I would recommend you switch over to https://github.com/Rackspace-Segment-Support/umb-protocol where the Avro schema is now maintained and pre-generated into Java. The MaaS, Salus, and UIM projects are all using umb-protocol.

itzg commented on src/test/java/com/rackspacecloud/metrics/ingestionservice/InfluxDBHelperTests.java:
```diff
@@ -31,143 +35,170 @@
                 tenantId, databaseName, rpName));
 
-        successfulIngestionTest(influxDBHelper, restTemplateMock, tenantId, databaseName, rpName);
+        successfulIngestionTest(influxDBHelper, restTemplateMock,
```
I don't see any assertions or Mockito verify calls. Is there some other way the unit tests are verifying success?

mritunjaykumar replied:
Added Mockito.verify()

itzg commented on src/test/java/com/rackspacecloud/metrics/ingestionservice/RollupIngestionServiceTests.java:
```diff
@@ -49,61 +49,62 @@
-        Assert.assertTrue(
-                String.format("Failed with batchProcessed count [%s]", batchProcessed), batchProcessed > 0);
-    }
+//	@Autowired
```
Did you still want this all commented out?

mritunjaykumar replied:
I will be working on rollup-ingestion next. That's why tests are commented out. It's temporary

itzg reviewed (commented):
Just a few things to consider, but nothing major.

Commit:
Incorporated following code review comments:
Comment: Can this class be combined with InfluxDBHelper since the suffixes "helper" and "utils" don't make it apparent the difference between these two?
Resolution: Done

Comment: I'm thinking you should wrap this in a synchronized(urlInfluxDBInstanceMap) block since I'm assuming calls to this originate from within a kafka consumer thread.
Resolution: I found two maps and converted both of them to ConcurrentHashMap

Comment: Doesn't seem to be needed since you declared a constructor.
Resolution: Changed to @AllArgsConstructor and removed constructor def from the code.

Comment: Could this be registered once in the constructor rather than doing a no-op register here?
Resolution: Moved it to constructor

Comment: Computing the toString of the records collection seems excessive when debug logs are not enabled. You could pass records directly and then let slf4j make the optimization here
Resolution: removed "stringification"

Comment: You can probably remove these content fixes since the maas-metrics-enrich is now normalizing those fields
	Removed it.
Resolution: I also removed String stringifiedData = new String(data) and pushed stringification into debug and error calls. So that stringification only happens in those cases rather than every time which is costly affair.

Comment: I would recommend you switch over to https://github.com/Rackspace-Segment-Support/umb-protocol where the Avro schema is now maintained and pre-generated into Java. The MaaS, Salus, and UIM projects are all using umb-protocol.
Resolution: Need to discuss this further.

Comment: I don't see any assertions or Mockito verify calls. Is there some other way the unit tests are verifying success?
Resolution: Added Mockito.verify()

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/InfluxDBHelper.java:
```diff
@@ -269,18 +267,16 @@ private boolean createDatabase(final String databaseName, final String baseUrl,
+        influxDB = this.influxDBFactory.getInfluxDB(instanceUrl);
+        influxDB.setLogLevel(InfluxDB.LogLevel.BASIC);
+        influxDB.enableBatch(batchOptions);
+        this.urlInfluxDBInstanceMap.put(instanceUrl, influxDB);
```
Using the `get` and `put` as separate calls loses the concurrency guarantees of `ConcurrentHashMap`. Instead use this

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfAbsent-K-java.util.function.Function-

which will atomically guarantee to either get the existing value or put the computed value if not present.

The race condition is discussed more here https://stackoverflow.com/a/14947844/121324

mritunjaykumar replied:
Done

itzg commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/utils/InfluxDBFactory.java:
```diff
@@ -0,0 +1,9 @@
+
+public class InfluxDBFactory {
+    public InfluxDB getInfluxDB(String url) {
+        return org.influxdb.InfluxDBFactory.connect(url);
```
I would have shifted all of https://github.com/racker/ceres-ingestion-service/pull/5/files/ccc7604d4a83f6829d3545c7586f766dd89e8b93..7d786becb258ed73dac913bc02f8fabb2d174b0f#diff-aff45cee99c0b2218cf7dc6301959b4fL269 into this method, which is actually what I did in the event engine ingest

https://github.com/racker/salus-event-engine-ingest/blob/master/src/main/java/com/rackspace/salus/event/ingest/services/KapacitorConnectionPool.java#L32

however, mine was simpler since I didn't yet need to tweak batch options.

mritunjaykumar replied:
Agree and done.

itzg commented on src/test/java/com/rackspacecloud/metrics/ingestionservice/InfluxDBHelperTests.java:
```diff
@@ -151,6 +152,9 @@ private void successfulIngestionTest(
 
         influxDBHelper.ingestToInfluxDb(payloadToIngestInInfluxDB, tenantId, measurement, rollupLevel);
+
+        verify(influxDBMock, Mockito.times(1))
```
Please have it verify the actual arguments instead of using `any()`. `any()` should only be used for `when` declarations.

Check out https://github.com/racker/salus-telemetry-monitor-management/blob/master/src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java if you want to see some various uses of verify and asserts.

mritunjaykumar replied:
Done

zack-shoylev commented on src/main/java/com/rackspacecloud/metrics/ingestionservice/listeners/UnifiedMetricsListener.java:
```diff
@@ -90,4 +68,16 @@ public void onIdleContainer(Map<TopicPartition, Long> map, ConsumerSeekCallback
+
+    public static String replaceSpecialCharacters(String inputString){
+        final String[] metaCharacters =
+                {"\\",":","^","$","{","}","[","]","(",")",".","*","+","?","|","<",">","-","&","%"," "};
```
Consider pre-compiling a regex pattern into a static variable, then reusing it when making a matcher to use with replaceAll, something like `pattern.matcher(text).replaceAll("_")` ; Pattern is also immutable, so this should remain thread-safe

zack-shoylev replied:
There's also StringUtils.containsAny but not sure if that would help

Commit:
Comment: Using the get and put as separate calls loses the concurrency guarantees of ConcurrentHashMap...
Resolution: Replaced two separate calls of `get` and `put` with `computeIfAbsent`

Comment: Please have it verify the actual arguments instead of using any(). any() should only be used for when declarations.
Resolution: Added actual arguments to verify call.

Commit:
Comment: I would have shifted all of https://github.com/racker/ceres-ingestion-service/pull/5/files/ccc7604d4a83f6829d3545c7586f766dd89e8b93..7d786becb258ed73dac913bc02f8fabb2d174b0f#diff-aff45cee99c0b2218cf7dc6301959b4fL269 into this method, which is actually what I did in the event engine ingest......

Resolution: I moved the code as suggested.

Commit:
Tests were failing because of last change, so fixed the test cases now. All are passing now.

mritunjaykumar commented:
All taken care of.

From: Geoff Bourne <notifications@github.com>
Reply-To: racker/ceres-ingestion-service <reply@reply.github.com>
Date: Friday, May 3, 2019 at 6:17 AM
To: racker/ceres-ingestion-service <ceres-ingestion-service@noreply.github.com>
Cc: Mritunjay Kumar <Mritunjay.Kumar@rackspace.com>, Mention <mention@noreply.github.com>
Subject: Re: [racker/ceres-ingestion-service] Jay/scale changes (#5)


@itzg commented on this pull request.

________________________________

In src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/InfluxDBHelper.java<https://github.com/racker/ceres-ingestion-service/pull/5#discussion_r280766135>:

>          BatchOptions batchOptions = BatchOptions.DEFAULTS;

         batchOptions.actions(this.numberOfPointsInAWriteBatch);

         batchOptions.flushDuration(this.writeFlushDurationMsLimit);

-//        batchOptions.bufferLimit(2000);

-

-        if(influxDB == null) {

-            influxDB = this.influxDBUtils.getInfluxDB(instanceUrl);

-            influxDB.setLogLevel(InfluxDB.LogLevel.BASIC);

-//            influxDB.enableBatch(BatchOptions.DEFAULTS);

-            influxDB.enableBatch(batchOptions);

-            this.urlInfluxDBInstanceMap.put(instanceUrl, influxDB);

-        }

+

+        influxDB = this.influxDBFactory.getInfluxDB(instanceUrl);

+        influxDB.setLogLevel(InfluxDB.LogLevel.BASIC);

+        influxDB.enableBatch(batchOptions);

+        this.urlInfluxDBInstanceMap.put(instanceUrl, influxDB);

Using the get and put as separate calls loses the concurrency guarantees of ConcurrentHashMap. Instead use this

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfAbsent-K-java.util.function.Function-

which will atomically guarantee to either get the existing value or put the computed value if not present.

The race condition is discussed more here https://stackoverflow.com/a/14947844/121324

________________________________

In src/main/java/com/rackspacecloud/metrics/ingestionservice/utils/InfluxDBFactory.java<https://github.com/racker/ceres-ingestion-service/pull/5#discussion_r280768240>:

> @@ -0,0 +1,9 @@

+package com.rackspacecloud.metrics.ingestionservice.utils;

+

+import org.influxdb.InfluxDB;

+

+public class InfluxDBFactory {

+    public InfluxDB getInfluxDB(String url) {

+        return org.influxdb.InfluxDBFactory.connect(url);

I would have shifted all of https://github.com/racker/ceres-ingestion-service/pull/5/files/ccc7604d4a83f6829d3545c7586f766dd89e8b93..7d786becb258ed73dac913bc02f8fabb2d174b0f#diff-aff45cee99c0b2218cf7dc6301959b4fL269 into this method, which is actually what I did in the event engine ingest

https://github.com/racker/salus-event-engine-ingest/blob/master/src/main/java/com/rackspace/salus/event/ingest/services/KapacitorConnectionPool.java#L32

however, mine was simpler since I didn't yet need to tweak batch options.

________________________________

In src/test/java/com/rackspacecloud/metrics/ingestionservice/InfluxDBHelperTests.java<https://github.com/racker/ceres-ingestion-service/pull/5#discussion_r280769270>:

> @@ -151,6 +152,9 @@ private void successfulIngestionTest(

         doNothing().when(influxDBMock).write(anyString(), anyString(), any(), any(), anyString());



         influxDBHelper.ingestToInfluxDb(payloadToIngestInInfluxDB, tenantId, measurement, rollupLevel);

+

+        verify(influxDBMock, Mockito.times(1))

Please have it verify the actual arguments instead of using any(). any() should only be used for when declarations.

Check out https://github.com/racker/salus-telemetry-monitor-management/blob/master/src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java if you want to see some various uses of verify and asserts.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/racker/ceres-ingestion-service/pull/5#pullrequestreview-233458628>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABYXN7PEKXJZXUEG6YO4OIDPTQ3MRANCNFSM4HEX2QNQ>.
